### PR TITLE
Improved Questlog

### PR DESCRIPTION
**Pull request by vladtepesch**
- completed quests are displayed
  - completely reworked quest log item drawing
  - spacing and positioning depends on list size
- fixed missing quest log caption
- selection follows cursor

as recently discussed here https://github.com/diasurgical/devilutionX/pull/2589#discussion_r688705609 
showing completed quests could be nice.

I implemented this as proposal. see the screenshots at the end.
the closed quests are not selectable. and also are skipped while selecting.

Two question that may be have to discussed:
- do we need an ini-option for showing solved quests?
- should we remove the "close quest log" line?
  This would bring some additional space and this panel is the only one in the game that does have a close btn

This PRs implementation is based (in termes of implementation style) on the original code. however there is an continuing/alternative version that may be the next step for cleaning the code: https://github.com/vladtepesch/devilutionX/tree/QuestLogImprovementsClass
It separates the questbook logic from the rest of questlogic in the `quest.cpp`. for keeping the interface constant it uses forward functions to the new `QuestBook` Singelton. The next logic step would be the removal if the `Questlog` functions of the `quest.h` and change the remaining code to use the `Questbook`

screen00
screen01
screen02
screen04
screen03
screen07
screen08

### CODE REVIEW

**Comment by AJenbo** [comment]
Instead of packing them tighter, would it be doable to make the list scroll instead?

**Comment by vladtepesch** [comment]
> 
> 
> Instead of packing them tighter, would it be doable to make the list scroll instead?

duable - sure, but I guess the changes would be much more extensive and not that locally anymore. I have not yet looked into the input handling part of the game. 

The last screenshot does contain the maximum number of quests in the game. so they all fit on the page. Its most unlikely that anyone gets a quest log that convoluted.  
May be this could be left for future improvement.

btw: wtf is is the "wandering trader"?

**Comment by qndel** [comment]
```cpp
	{      17,          -1, DTYPE_NONE,      Q_GIRL,      100,    SL_NONE,         true,               TEXT_GIRL2,    N_("Little Girl")              },
	{      19,          -1, DTYPE_NONE,      Q_TRADER,    100,    SL_NONE,         true,               TEXT_TRADER,   N_("Wandering Trader")         },
	{      17,          17, DTYPE_NONE,      Q_DEFILER,   100,    SL_NONE,         false,              TEXT_DEFILER1, N_("The Defiler")              },
	```
	
trader is an unused quest

**Comment by FitzRoyX** [comment]
> Instead of packing them tighter, would it be doable to make the list scroll instead?

This is part of why I split the quests up into 2 pages. It allows the entries to be somewhat spaced out (prevent misclicks), and it allows "expansion" quests (hellfire and mod) to be differentiated from core ones. That it avoids a scrollbar too is just a side effect.

Since no other windows have a "close" feature, I felt that the "close quest" entry was expendable and I used that space for flipping pages.

Still, I'm very grateful of this PR even if I don't get my exact design.

**Comment by julealgon** [comment]
> do we need an ini-option for showing solved quests?

Absolutely not, IMHO. This is a simple net improvement and there is zero reason to go back.

> should we remove the "close quest log" line?
> This would bring some additional space and this panel is the only one in the game that does have a close btn

This is a fair question regardless of spacing. Considering other panels do not have close buttons (inventory, character screen) and that we have a "QUESTS" button in the UI for toggling this panel, I think it would make sense and be consistent to remove the "close quest log" line.

I'm not a fan of the alphabetical ordering, however. To me it should either be in the "canonical order" (i.e. the order that they appear in-game) _or_ the order that the player finds them.

Lastly, we could also consider a tabbed design here: one tab for pending quests, and another for completed ones. Just throwing this out there (I don't necessarily love the idea myself, but might serve as food for thought for others).

**Comment by vladtepesch** [comment]
> 
> 
> I'm not a fan of the alphabetical ordering, however. To me it should either be in the "canonical order" (i.e. the order that they appear in-game) _or_ the order that the player finds them.

order of collecting would require saving this order --> I would vote against this.

But we could setup an arbitrary order in which we want to display them. The status quo was the order of the quest array that has no logical ordering in any sense. 

Feel free to post a ordering (best using quest defines from the game). I will put it in.

> Lastly, we could also consider a tabbed design here: one tab for pending quests, and another for completed ones. Just throwing this out there (I don't necessarily love the idea myself, but might serve as food for thought for others).

I also thought about this but at the end this would have the same problem of space - because at the end everything is at the completed page and you have to toggle between them.  Thatswhy I dropped this idea.
And currently without any mods the number of all quests fits on one page. And the last screenshot already shows more quests than you normally could collect.  
So if we drop the close-point everything is fine in any case.

**Comment by julealgon** [comment]
> order of collecting would require saving this order --> I would vote against this.

True, it would be more complex. That shouldn't be the only reason to _not_ do it, though. We should focus on the desired UX. If showing based on find date, we should try to collect that at some point (necessarily as part of this effort).

> Feel free to post a ordering (best using quest defines from the game). I will put it in.

I personally don't have the main ordering in hand right now. Someone else can help here?

> I also thought about this but at the end this would have the same problem of space - because at the end everything is at the completed page and you have to toggle between them. Thatswhy I dropped this idea.

Oh for sure. I wasn't putting the idea on the table for space conservation purposes, but for organization/filtering purposes instead. Usually, users would want to focus on pending quests, so we could have completed ones not show in the "main" tab.

You definitely still need a solution for the spacing issue even with the tabs in place.

**Comment by vladtepesch** [comment]
> I personally don't have the main ordering in hand right now. Someone else can help here?
I throw a look at this:
maybe just use the single player level.

But I am not sure about the helfire towns people  - putting them inbetween or right at the end ?

```
quest          _qdlvl
=======================
Q_BUTCHER,      2
Q_PWATER,       2
Q_SKELKING,     3
Q_GARBUD,       4
Q_LTBANNER,     4
Q_ROCK,         5
Q_BLOOD,        5
Q_SCHAMB,       6
Q_BLIND,        7
Q_ZHAR,         8
Q_MUSHROOM,     9
Q_ANVIL,       10
Q_WARLORD,     13
Q_VEIL,        14
Q_BETRAYER,    15
Q_DIABLO,      15

Q_FARMER,       9
Q_JERSEY,       9
Q_GIRL,        17

Q_GRAVE,       17
Q_DEFILER,     17
Q_TRADER,      19
Q_CORNSTN,     21
Q_NAKRUL,      21
```

**Comment by AJenbo** [comment]
I think I prefer them in two groupes and canonical order

**Comment by vladtepesch** [comment]
that is the canonical order? and that do you mean with two groups? do you want an additional separation between Diablo and HF quests? this would imply that there will be 4 groups. 2 active and 2 completed.

**Comment by AJenbo** [comment]
Erh, by two groups I'm implying 2, not 4. Active and completed.

**Comment by julealgon** [comment]
> additional separation between Diablo and HF

My two cents is that this should _never_ happen in a customer-facing way about anything in the game. 

Categorizing anything based on Diablo/Hellfire is immersion breaking and a form of "UI special casing". We should avoid all forms of special casing as much as possible.

**Comment by vladtepesch** [comment]
okay - good, I am thinking the same. 

Then I just misinterpreted AJenbos answer because I never questioned the working/complete grouping but the quest order within these groups

**Comment by FitzRoyX** [comment]
vlad, make sure your code accounts for different vertical resolutions. I'm at 720p and the text is where 480p would be.

**Comment by vladtepesch** [comment]
I'll have to check this. But I guess, since i oriented on the original code, also the original quest log would draw the options at strange places. Can you post a screenshot so that i can verify that I can create settings that reflect your problems? (maybe also the relevant part of your ini-file)

**Comment by FitzRoyX** [comment]
Untitled-1f

**Comment by vladtepesch** [comment]
ahh - I see. I guess I take the panel position into account twice. one time in the `PrintQLString` and one time at the `DrawQuestLog`.

thank you for pointing this out. I'll fix this in the evening.

**Comment by vladtepesch** [comment]
> 
> 
> vlad, make sure your code accounts for different vertical resolutions. I'm at 720p and the text is where 480p would be.

should be corrected

**Comment by FitzRoyX** [comment]
Great, thanks. Per my screenshot, the main problem now is sorting the groups so that there's a sense of chronology to events. I didn't meet Lazarus before Gharbad, etc.

**Comment by vladtepesch** [comment]
> 
> 
> Great, thanks. Per my screenshot, the main problem now is sorting the groups so that there's a sense of chronology to events. I didn't meet Lazarus before Gharbad, etc.

yeah I am waiting for someone to post a order or to answer about mine ;)

Directly using the `_qdlevel` would be one option (allthough I would making the order explicit becasue of the ambiguity) but the hive starting quest appear in the midde of the other. 
My usual order is playing 1-16 and then doing the Hive and Crypt.

But I guess the intended order would be to do the Hf levels inbetween with doing the hell levels last.

**Comment by vladtepesch** [comment]
so maybe this would be fine:

```
quest          _qdlvl
=======================
Q_BUTCHER,      2
Q_PWATER,       2
Q_SKELKING,     3
Q_GARBUD,       4
Q_LTBANNER,     4
Q_ROCK,         5
Q_BLOOD,        5
Q_SCHAMB,       6
Q_BLIND,        7
Q_ZHAR,         8
Q_MUSHROOM,     9
Q_ANVIL,       10
Q_FARMER,       9
Q_JERSEY,       9
Q_GIRL,        17
Q_GRAVE,       17
Q_DEFILER,     17
Q_TRADER,      19
Q_CORNSTN,     21
Q_NAKRUL,      21
Q_WARLORD,     13
Q_VEIL,        14
Q_BETRAYER,    15
Q_DIABLO,      15
```

**Comment by julealgon** [comment]
Here is my suggestion. 
- Hell and Crypt are mostly the same level, as are caves and hive.
- Grave should appear after defiler, so I swapped them.
- Also moved nakrul before diablo
- Farmer/Jersey should be accessible as soon as hive is, so sooner than mushroom which requires you to get the book inside caves already.
- Warlord and cornerstone are mostly equivalent in level imho too. I put the cornerstone before warlord somewhat arbitrarily.

| quest | _qdlvl |
| --- | --- |
| Q_BUTCHER | 2 |
| Q_PWATER|2|
| Q_SKELKING|3|
| Q_GARBUD|4|
| Q_LTBANNER|4 |
| Q_ROCK|5|
| Q_BLOOD|5|
| Q_SCHAMB|6|
| Q_BLIND|7|
| Q_ZHAR|8|
| Q_FARMER|9|
| Q_JERSEY|9|
| Q_MUSHROOM|9|
| Q_ANVIL|10|
| Q_GIRL|17|
| Q_DEFILER|17|
| Q_GRAVE|17|
| Q_TRADER|19|
| Q_WARLORD|13|
| Q_CORNSTN|21|
| Q_VEIL|14|
| Q_BETRAYER|15|
| Q_NAKRUL|21|
| Q_DIABLO|15|

**Comment by vladtepesch** [comment]
Defiler/grave was my fault.  
But why putting Nakrul at the end? the quest is opened shortly after entering the grave.

and putting warlord in the middle of the grave quests does not fit for me. because most likely you finish either the hell or the grave first.

Also I would put the farmer/jersey after the finishing the caves -> after anvil and fungus because most likely a player does (or at least I would) not stop in the middle of caves because then I would lose my townportal and would have to walk down again)

**Comment by julealgon** [comment]
> But why putting Nakrul at the end? the quest is opened shortly after entering the grave.

It is still after the grave in the "hellfire timeline" there. I see these quests as two branching paths at 2 points in the history, but their overall order should be combined. This is exactly to avoid the division between Diablo and Hellfire and make the quests mesh more naturally.
 
> and putting warlord in the middle of the grave quests does not fit for me. because most likely you finish either the hell or the grave first.

Again, see the branching paths argument above. Yes, most players would follow one or the other, and then they would see the quests sequentially in that branch. I think that's fine.

> Also I would put the farmer/jersey after the finishing the caves -> after anvil and fungus because most likely a player does (or at least I would) not stop in the middle of caves because then I would lose my townportal and would have to walk down again)

Nest shares the same level/difficulty with caves, so they are parallel. Having nest totally after caves would be bad IMHO.

**Comment by vladtepesch** [comment]
> Again, see the branching paths argument above. Yes, most players would follow one or the other, and then they would see the quests sequentially in that branch. I think that's fine.

But exact that branching is my argument to not mesh them.

because most probable you would follow either the one branch completely before entering the other branch.  so presenting the quests at a block for each branch would prevent meshing them.  
I do not see the logic to mesh the quests of two parallel branches because mostly you would not mesh the solving of the levels because that would involve a lot of backtracking because of lost town portals

**Comment by FitzRoyX** [comment]
Hellfire wasn't planned and the hero sticks the soulstone in his head, so they just shoehorned their own story arc into the middle of the game. It doesn't try hard enough to be part of the original story to justify mixing them IMO. So I'd divide them, but this also means that the grouping of active and completed has to go.

**Comment by Trihedraf** [comment]
What happens when you are only playing  Diablo and not hellfire? Because I never play with hellfire so I wouldn't want the hellfire quests to be showing up.

**Comment by vladtepesch** [comment]
> 
> 
> What happens when you are only playing Diablo and not hellfire? Because I never play with hellfire so I wouldn't want the hellfire quests to be showing up.

only quests you already started or completed are shown in questlog

File: Source/quests.cpp [hunk 1]
```diff
@@ -833,14 +833,18 @@ void StartQuestlog()
 		}
 	}
 
-	selectedEntry = 0;
+	if (fistfinishedEntry != 0) {
+		selectedEntry = 0;
+	} else {
+		selectedEntry = qlistCnt;
+	}
```
Comment (qndel) [inline Source/quests.cpp]:
```suggestion
	selectedEntry = (fistfinishedEntry == 0 ? qlistCnt : 0);
```
also you have a typo, should be firstFinishedEntry, not fistfinishedEntry

Comment (vladtepesch) [reply Source/quests.cpp]: oh - you are right - thanks to auto completion, I never noticed ;)

File: Source/quests.cpp [hunk 2]
```diff
@@ -72,7 +72,7 @@ int WaterDone;
 
 int qlist[MAXQUESTS];  ///< indices of quests to display in quest log window. `fistfinishedEntry` are active quests the rest are completed
 int qlistCnt;          ///< overall number of qlist entries
-int fistfinishedEntry; ///< first (nonselectable) finished quest in list
```
Comment (qndel) [inline Source/quests.cpp]: image

Comment (vladtepesch) [reply Source/quests.cpp]: omg - and I thought you ment the casing 🤣

**Comment by vladtepesch** [comment]
About the order topic:
this is the current ordering, I implemented: 
I fixed the one ordering bug that @julealgon pointed out, but (for now) rejected the meshing of  parallel paths.


```
_qdlvl,  _qdtype,    bookOrder, 
     2,  Q_BUTCHER,      0,     
     2,  Q_PWATER,       1,     
     3,  Q_SKELKING,     2,     
     4,  Q_GARBUD,       3,     
     4,  Q_LTBANNER,     4,     
     5,  Q_ROCK,         5,     
     5,  Q_BLOOD,        6,     
     6,  Q_SCHAMB,       7,     
     7,  Q_BLIND,        8,     
     8,  Q_ZHAR,         9,     
     9,  Q_MUSHROOM,    10,     
    10,  Q_ANVIL,       11,     
     9,  Q_FARMER,      12,     
     9,  Q_JERSEY,      13,     
    17,  Q_GIRL,        14,     
    17,  Q_DEFILER,     15,     
    19,  Q_TRADER,      16,     
    17,  Q_GRAVE,       17,     
    21,  Q_CORNSTN,     18,     
    21,  Q_NAKRUL,      19,     
    13,  Q_WARLORD,     20,     
    14,  Q_VEIL,        21,     
    15,  Q_BETRAYER,    22,     
    15,  Q_DIABLO,      23,     
```

so may be, we can make a small poll:
 👍 - I am fine with it  
 🚀 - move hive quests above cave
 🎉 - meshing them together  
 👎 - I want something completely different

**Comment by vladtepesch** [comment]
rebased and tidied up commits

**Comment by vladtepesch** [comment]
so maybe we can get this merged to avoid further conflicts?

**Comment by AJenbo** [comment]
That's also true for all the other PRs i haven't had time to merge yet :(

I would be more prone to merge it if you can figure out some sort of scrolling, doesn't need a scrollbar. Also you can drop the title and close item, we already have esc, space and the UI button and it's none of the other panels.

**Comment by vladtepesch** [comment]
I do not think that we need the scrolling.  
The maximum amount of quests (23 - if farmer and cow was done) does not look that tightly packed. And if we drop the close and the title line, then there is an additional pixel for each line.
I also would not know how to cosistently implement it. The completed quests are not selectable. But without beeing selectable - how would you scroll over them. And if selectable, it would be strange if they are not activatable. And if also activatable, which entry should be played?

Also I am totally against scrolling without scroll bars. The UI should have an indication that something is scrollable.

btw: was does happen in original game if more quests are open than the amount that will fit the screen with this generous spacing used originally?

**Comment by AJenbo** [comment]
You can use the "give quest #" command to find out now :)

A bit more
image

to much:
image

Lots more
image

**Comment by AJenbo** [comment]
Remove the close button and title and we have a deal :)

**Comment by vladtepesch** [comment]
@julealgon why the confused smiley? is it just because of the ordering?

**Comment by vladtepesch** [comment]
here is an depiction of all variants (the last wo rows (23 and 24) quests would be rather unusual)

File: Source/quests.cpp [hunk 3]
```diff
@@ -38,42 +38,50 @@ int ReturnLevel;
...
-int qtopline;
-int qline;
-quest_id qlist[MAXQUESTS];
-int numqlines;
 int WaterDone;
 
+quest_id qlist[MAXQUESTS]; ///< indices of quests to display in quest log window. `fistfinishedEntry` are active quests the rest are completed
+int qlistCnt;              ///< overall number of qlist entries
+int firstFinishedEntry;    ///< first (nonselectable) finished quest in list
+int selectedEntry;         ///< currently selected quest list item
```
Comment (AJenbo) [inline Source/quests.cpp:76]:
```suggestion
/** Indices of quests to display in quest log window. `fistfinishedEntry` are active quests the rest are completed */
quest_id qlist[MAXQUESTS];
/** Overall number of qlist entries */
int qlistCnt;
/** First (nonselectable) finished quest in list */
int firstFinishedEntry;
/** Currently selected quest list item */
int selectedEntry;
```

File: Source/quests.cpp [hunk 4]
```diff
@@ -721,59 +745,95 @@ void ResyncQuests()
...
+		}
+	}
+
+	std::sort(&qlist[0], &qlist[firstFinishedEntry], sortQuestIdx);
+	std::sort(&qlist[firstFinishedEntry], &qlist[qlistCnt], sortQuestIdx);
+
+	bool twoBlocks = (firstFinishedEntry != 0) && (firstFinishedEntry < qlistCnt);
+
+	topY = 0;
+	act2finSpacing = (!twoBlocks) ? 0 : lineHeight / 2;
```
Comment (AJenbo) [inline Source/quests.cpp:792]: We don't put parenthesis around evaluations, instead we require them to be explicitly boolean.
```suggestion
	act2finSpacing = !twoBlocks ? 0 : lineHeight / 2;
```

This is also the case in a few other places but not a bit problem, just a bit of style differences.

File: Source/quests.cpp [hunk 5]
```diff
@@ -721,59 +745,95 @@ void ResyncQuests()
...
+
+	int overallMinHeight = qlistCnt * lineHeight + act2finSpacing;
+	int space = panelInnerRect.size.height;
+
+	if (qlistCnt > 0) {
+		int additionalSpace = (space - overallMinHeight);
+		int addLineSpacing = additionalSpace / (qlistCnt);
```
Comment (AJenbo) [inline Source/quests.cpp:799]: the parenthesis feels a bit inconsistent:

```suggestion
		int additionalSpace = space - overallMinHeight;
		int addLineSpacing = additionalSpace / qlistCnt;
```

File: Source/quests.cpp [hunk 6]
```diff
@@ -721,59 +745,95 @@ void ResyncQuests()
...
+		}
+
+		int overallHeight = qlistCnt * lineSpacing + act2finSpacing;
+		topY += (space - overallHeight) / 2;
 	}
-	qline = 22;
-	if (numqlines != 0)
-		qline = qtopline;
+
+	selectedEntry = (firstFinishedEntry == 0) ? -1 : 0;
```
Comment (AJenbo) [inline Source/quests.cpp:812]:
```suggestion
	selectedEntry = firstFinishedEntry == 0 ? -1 : 0;
```

File: Source/quests.cpp [hunk 7]
```diff
@@ -721,59 +745,95 @@ void ResyncQuests()
...
+		if (quest._qactive == QUEST_DONE || quest._qactive == QUEST_HIVE_DONE) {
+			qlist[qlistCnt] = quest._qidx;
+			qlistCnt++;
+		}
+	}
+
+	std::sort(&qlist[0], &qlist[firstFinishedEntry], sortQuestIdx);
+	std::sort(&qlist[firstFinishedEntry], &qlist[qlistCnt], sortQuestIdx);
+
+	bool twoBlocks = (firstFinishedEntry != 0) && (firstFinishedEntry < qlistCnt);
```
Comment (AJenbo) [inline Source/quests.cpp:789]:
```suggestion
	bool twoBlocks = firstFinishedEntry != 0 && firstFinishedEntry < qlistCnt;
```

**Comment by AJenbo** [comment]
> here is an depiction of all variants regarding number of quests and nosolved/some solved/all solved (the last wo rows (23 and 24) quests would be rather unusual)
> 

Looks good 👍

**Review by AJenbo — APPROVED** [review]

File: Source/quests.cpp (hunk 3 above)
Comment (vladtepesch) [reply Source/quests.cpp]: that was doxygen

Comment (vladtepesch) [reply Source/quests.cpp]: (one of the variants, you can put after the members, you want to document)

**Comment by vladtepesch** [comment]
great, thanks for merging. did you took a look on the other branch, I mentioned at beginning?

would you find this kind of refactoring desireable? (the code for the questbook is extracted into a own class+file and uses a pimpl pattern to hide the internals and dependencies)
If yes, then I rebase the changes onto the current implementation

**Comment by AJenbo** [comment]
Sorry I missed that, what branch are you referring to? What you describe sounds interesting, but not something we have applied anywhere else so would probably have to evaluate it a bit.

**Comment by vladtepesch** [comment]
I mentioned this in the opening post. the proof of concept implementation handels that change transparently by just forwarding the original funcitions to the new to avoid changing this all over.

**Comment by AJenbo** [comment]
Ok, must have missed it. Will take a look later and give you some feedback.

**Comment by vladtepesch** [comment]
Maybe the separation of the interface and implementation class (pimpl) is not that needed for this project since there is no need to keep an library API binary compatible. So maybe the Implementation class could be directly exposed in header.

File: Source/quests.cpp (hunk 3 above)
Comment (AJenbo) [reply Source/quests.cpp]: Ok, it's not the variant we use in the rest of the project

**Pull request merged**